lib/json.go

```diff
@@ -213,10 +218,34 @@ func encodeJSON(val ref.Val) ref.Val {
 		return types.NewErr("failed to get native value for JSON")
 	}
 	b, err := json.Marshal(v)
-	if err != nil {
+	errType := &json.UnsupportedTypeError{}
+	switch {
+	case err == nil:
+		return types.String(b)
+	case errors.As(err, &errType):
+		// We could not marshal over ref.Val most likely, so convert
+		// to protobuf as an intermediate and retry.
+		v, err := val.ConvertToNative(reflect.TypeOf(&structpb.Value{}))
+		if err != nil {
+			return types.NewErr("failed proto conversion: %v", err)
+		}
+		b, err := protojson.MarshalOptions{}.Marshal(v.(proto.Message))
+		if err != nil {
+			return types.NewErr("failed native conversion: %v", err)
+		}
+		var res interface{}
+		err = json.Unmarshal(b, &res)
+		if err != nil {
+			return types.NewErr("failed json conversion: %v", err)
+		}
+		b, err = json.Marshal(res)
+		if err != nil {
+			return types.NewErr("failed to marshal value to JSON: %v", err)
+		}
```

Comment:
We could omit these lines, since protojson serialises to JSON, [but](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson#MarshalOptions.Marshal)

> Marshal marshals the given proto.Message in the JSON format using options in MarshalOptions. **Do not depend on the output being stable. It may change over time across different versions of the program.**

Reply:
Better to be safe.

For my own reference I was tracing through the code to see what happens in the new test case. These were the transformations:

`*types.baseMap -> ConvertToNative() -> *structpb.Value -> protojson.Marshal() -> JSON as []uint8 -> json.Unmarshal() -> map[string]any -> json.Marshal() -> string`

Reply:
Yes, not this nicest path.